services/telemetry/config.go

```diff
@@ -0,0 +1,313 @@
+	}
+	c.Reporting.SendTimeout = sendTimeout
+
+	if d, err := time.ParseDuration(os.Getenv(c.Reporting.IntervalEnv)); err == nil && d > 0 {
+		l.Warnf("Interval changed to %s.", d)
+		c.Reporting.Interval = d
+	}
+	if d, err := time.ParseDuration(os.Getenv(c.Reporting.RetryBackoffEnv)); err == nil && d > 0 {
+		l.Warnf("Retry backoff changed to %s.", d)
+		c.Reporting.RetryBackoff = d
+	}
+
+	ds := c.DataSources
+
+	vmdb := ds.VM
+	if vmdb.Enabled {
+		if vmdb.TimeoutStr != "" {
+			timeout, err := time.ParseDuration(vmdb.TimeoutStr)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse duration [%s]", ds.VM.Timeout)
+			}
+			vmdb.Timeout = timeout
+		}
+	}
+
+	qandb := ds.QanDBSelect
+	if qandb.Enabled {
+		if qandb.TimeoutStr != "" {
+			timeout, err := time.ParseDuration(qandb.TimeoutStr)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse duration [%s]", ds.QanDBSelect.Timeout)
+			}
+			qandb.Timeout = timeout
+		}
+	}
+
+	pmmdb := ds.PmmDBSelect
+	if pmmdb.Enabled {
+		if pmmdb.TimeoutStr != "" {
+			timeout, err := time.ParseDuration(pmmdb.TimeoutStr)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse duration [%s]", ds.PmmDBSelect.Timeout)
+			}
+			pmmdb.Timeout = timeout
+		}
+	}
+
+	return nil
+}
+
+func (c *ServiceConfig) loadConfig(location string) ([]Config, error) { //nolint:cyclop
+	matches, err := filepath.Glob(location)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	var fileConfigs []FileConfig //nolint:prealloc
+	var fileCfg FileConfig
+	for _, match := range matches {
+		buf, err := ioutil.ReadFile(match) //nolint:gosec
+		if err != nil {
+			return nil, errors.Wrapf(err, "error while reading config [%s]", match)
+		}
+		if !c.DisableSigningVerification {
+			bufSign, err := ioutil.ReadFile(match + ".minisig") //nolint:gosec
+			if err != nil {
+				return nil, errors.Wrapf(err, "error while reading config [%s]", match)
+			}
+			valid := false
+			for _, publicKey := range c.Signing.trustedPublicKeysParsed {
+				if ok := minisign.Verify(publicKey, buf, bufSign); ok {
+					valid = true
+					break
+				}
+			}
+			if !valid {
+				return nil, errors.Errorf("signature verification failed for [%s]", match)
+			}
+		}
+		if err := yaml.Unmarshal(buf, &fileCfg); err != nil {
+			return nil, errors.Wrapf(err, "cannot unmashal config [%s]", match)
+		}
+		fileConfigs = append(fileConfigs, fileCfg)
+	}
+
+	if c.LoadDefaults {
+		defaultConfigBytes := []byte(defaultConfig)
+		if err := yaml.Unmarshal(defaultConfigBytes, &fileCfg); err != nil {
+			return nil, errors.Wrap(err, "cannot unmashal default config")
+		}
+		fileConfigs = append(fileConfigs, fileCfg)
+	}
+
+	if err := c.validateConfig(fileConfigs); err != nil {
+		c.l.Errorf(err.Error())
+	}
+
+	return c.merge(fileConfigs), nil
+}
+
+func (c *ServiceConfig) merge(cfgs []FileConfig) []Config {
+	var result []Config
+	ids := make(map[string]bool)
```

Comment:
Don't agree, with bool a bit more clear. Simple problem - simple solution.

if I'd use map[string]struct{}
setting a value will look like

```
ids := make(map[string]struct{})
ids["<some-id>"] = struct{}{}
```

compare that to 

```
ids := make(map[string]bool)
ids["<some-id>"] = true
```

Of course we can create a set Set type alias and create methods that hide all this mechanics. Again, simple problem - simple solution.

Reply:
Using empty structs is idiomatic way to create sets in go. In your code you never check actual value, so it will works the same even if you will assign `false`. Main problem with this approach that at first sight its no clear is values in the map are important or not (i.e. is it set or map of bools).

Reply:
What you are describing is not an objective statement. But rather a personal opinion, I do agree that value here is not that important. And it is not being used. We can use `map[string]*struct{}` so that we can set `nil` as a value. But it is also not an ideal solution either, because we risk having null dereference error. Plus, it doesn't contribute to clarity.

Again, I don't want to spend days and make another 5 lines to be ideal to everyone.

Reply:
Here is another benefit of using empty structs for sets: That's how we do in PMM.